src/exp/list_exp.rs

```diff
@@ -0,0 +1,752 @@
+        let args = vec![
+            ExpressionArgument::Value(Value::from(ListExpOp::RemoveByValue as i64)),
+            ExpressionArgument::Value(Value::from(ListReturnType::None as u8)),
+            ExpressionArgument::FilterExpression(value),
+            ExpressionArgument::Context(ctx.to_vec()),
+        ];
+        ListExpression::add_write(bin, ctx, args)
+    }
+
+    /// Create expression that removes list items identified by values.
+    pub fn remove_by_value_list(
+        values: FilterExpression,
+        bin: FilterExpression,
+        ctx: &[CdtContext],
+    ) -> FilterExpression {
+        let args = vec![
+            ExpressionArgument::Value(Value::from(ListExpOp::RemoveByValueList as i64)),
+            ExpressionArgument::Value(Value::from(ListReturnType::None as u8)),
+            ExpressionArgument::FilterExpression(values),
+            ExpressionArgument::Context(ctx.to_vec()),
+        ];
+        ListExpression::add_write(bin, ctx, args)
+    }
+
+    /// Create expression that removes list items identified by value range (valueBegin inclusive, valueEnd exclusive).
+    /// If valueBegin is null, the range is less than valueEnd. If valueEnd is null, the range is
+    /// greater than equal to valueBegin.
+    pub fn remove_by_value_range(
+        value_begin: Option<FilterExpression>,
+        value_end: Option<FilterExpression>,
+        bin: FilterExpression,
+        ctx: &[CdtContext],
+    ) -> FilterExpression {
+        let mut args = vec![
+            ExpressionArgument::Context(ctx.to_vec()),
+            ExpressionArgument::Value(Value::from(ListExpOp::RemoveByValueInterval as i64)),
+            ExpressionArgument::Value(Value::from(ListReturnType::None as u8)),
+        ];
+        if let Some(val_beg) = value_begin {
+            args.push(ExpressionArgument::FilterExpression(val_beg));
+        } else {
+            args.push(ExpressionArgument::FilterExpression(Expression::nil()));
+        }
+        if let Some(val_end) = value_end {
+            args.push(ExpressionArgument::FilterExpression(val_end));
+        }
+        ListExpression::add_write(bin, ctx, args)
+    }
+
+    /// Create expression that removes list items nearest to value and greater by relative rank.
+    ///
+    /// ```
+    /// // Examples for ordered list [0,4,5,9,11,15]:
+    /// // (value,rank) = [removed items]
+    /// // (5,0) = [5,9,11,15]
+    /// // (5,1) = [9,11,15]
+    /// // (5,-1) = [4,5,9,11,15]
+    /// // (3,0) = [4,5,9,11,15]
+    /// // (3,3) = [11,15]
+    /// // (3,-3) = [0,4,5,9,11,15]
+    /// ```
+    pub fn remove_by_value_relative_rank_range(
+        value: FilterExpression,
+        rank: FilterExpression,
+        bin: FilterExpression,
+        ctx: &[CdtContext],
+    ) -> FilterExpression {
+        let args = vec![
+            ExpressionArgument::Value(Value::from(ListExpOp::RemoveByValueRelRankRange as i64)),
+            ExpressionArgument::Value(Value::from(ListReturnType::None as u8)),
```

Comment:
Why is the return type hard-coded to `None`? Shouldn't we allow the user to decide whether they want any return values? Same for the other `remove_by_*` functions below.

Reply:
Good question. Im not sure why its hardcoded. Thats the way that the Java Client does it. I tried to keep as close as possible.

Reply:
@kportertx Any input on this? Does the server support returning data for the list "remove-by" expressions? If not, why do the clients still need to pass the return type argument?

Reply:
For modify operations in the expression system we always return the modified Bin to other operations. Some operations support returning a result bin in a modify operation, this result bin is discarded in the expression context.

The argument here is an argument intrinsic to the list operation (not the expression system itself). Since the expression system cannot use the result-bin anyhow, the client abstracts the argument away from the user.